### PR TITLE
Dlo / SLO support. fixes #16

### LINKED ISSUE #16: support slo/dlo
See if Gleb's middlewares can help any.

https://review.openstack.org/#/c/287157/7/go/proxyserver/slo.go
https://review.openstack.org/#/c/287157/7/go/proxyserver/dlo.go

### PR DESCRIPTION
**Pull request by dpgoetz**

### CODE REVIEW

File: proxyserver/middleware/largeobject.go [hunk 1]
```diff
@@ -0,0 +1,755 @@
...
+	allowWrite       bool
+	xloFuncName      string
+}
+
+func (sw *segWriter) WriteUpstreamHeader() {
+	if sw.isSlo {
+		sw.Header().Set("X-Static-Large-Object", "True")
+	}
+	sw.ResponseWriter.WriteHeader(sw.Status)
+	sw.Header().Set("X-Static-Large-Object", "")
```
Comment (corystone) [inline proxyserver/middleware/largeobject.go]: Is this any different than Del()?

File: proxyserver/middleware/largeobject.go [hunk 2]
```diff
@@ -0,0 +1,755 @@
...
+}
+
+func (sw *segWriter) WriteHeader(status int) {
+	sw.Status = status
+	if sw.throwAwayHeader { // DLO segments cannot be sub-DLOs. just treat as bytes
+		return
+	}
+	cLen, _ := strconv.ParseInt(sw.Header().Get("Content-Length"), 10, 64)
+	sw.ContentLength = int64(cLen)
```
Comment (corystone) [inline proxyserver/middleware/largeobject.go]: This is already int64.

File: proxyserver/middleware/largeobject.go [hunk 3]
```diff
@@ -0,0 +1,755 @@
...
+	sw.manifestBytes = bytes.NewBuffer(make([]byte, 0))
+	if theDlo := sw.Header().Get("X-Object-Manifest"); theDlo != "" && sw.xloFuncName != "get" {
+		sw.isDlo = true
+		sw.dloHeader = theDlo
+		sw.allowWrite = false // because DLO manifests can be segments- do not write out bytes yet
+	}
+	if isSlo := sw.Header().Get("X-Static-Large-Object"); isSlo == "True" {
+		sw.isSlo = true
+		sw.cacheBytes = true
+		sw.Header().Set("X-Static-Large-Object", "")
```
Comment (corystone) [inline proxyserver/middleware/largeobject.go]: Same question about Del().

File: proxyserver/middleware/largeobject.go [hunk 4]
```diff
@@ -0,0 +1,755 @@
...
+	if sw.isSlo {
+		sw.Header().Set("X-Static-Large-Object", "True")
+	}
+	sw.ResponseWriter.WriteHeader(sw.Status)
+	sw.Header().Set("X-Static-Large-Object", "")
+}
+
+func (sw *segWriter) WriteHeader(status int) {
+	sw.Status = status
+	if sw.throwAwayHeader { // DLO segments cannot be sub-DLOs. just treat as bytes
```
Comment (corystone) [inline proxyserver/middleware/largeobject.go]: If this happens, do you care about the status? Can it be an error?

Comment (dpgoetz) [inline proxyserver/middleware/largeobject.go]: ya- you want to know if it is a 404. ln. 232

**Pull request merged**